feat(DocumentMergeService): Add missing actors field for file merge, …

AxelDeneu opened the pull request:
Hello,

I noticed that some actor-related data was missing, so I’ve added the missing pieces and also took the opportunity to clean up the `DocumentMergeService`.

I also renamed the `payer()` method to `payor()` which fits more the naming convention in the project.

While reviewing the code, I realized that the `DocumentMergeService` could potentially benefit from further refactoring. Specifically, I suggest splitting it into two separate services:

- One dedicated to data collection.
- Another focused on template processing.

This would help improve modularity, readability, and maintainability.

Let me know your thoughts or if you have suggestions for other improvements!

Thanks!

AxelDeneu committed:
feat(DocumentMergeService): Add missing actors field for file merge, improved readibility in the class

jjdejong commented on app/Models/Matter.php:
```diff
@@ -150,6 +150,32 @@ public function agents()
             ->wherePivot('role', 'AGT');
     }
 
+    public function sharedAgents()
```
There shouldn't be more than one agent, yet this is not checked, so the user could add multiple agents, among which only one should be returned.

AxelDeneu replied:
I wasn't sure if this was the case, I'll apply a fix for this.
Also, commenting each method will be better.

jjdejong replied:
Same for agents() of course. I forgot to react on this in the previous PR, sorry.

jjdejong commented on app/Models/Matter.php:
```diff
@@ -150,6 +150,32 @@ public function agents()
+            ->wherePivot('shared', 1);
+    }
+
+    public function secondaryAgents()
```
There shouldn't be more than one, although the user can add multiple agents.

jjdejong commented on app/Models/Matter.php:
```diff
@@ -150,6 +150,32 @@ public function agents()
+            ->wherePivot('role', 'AGT2');
+    }
+
+    public function sharedSecondaryAgents()
```
There shouldn't be more than one, although the user can add multiple agents.

jjdejong commented on app/Models/Matter.php:
```diff
@@ -150,6 +150,32 @@ public function agents()
+            ->wherePivot('shared', 1);
+    }
+
     public function writers()
```
There shouldn't be more than one, although the user can add multiple.

jjdejong commented on app/Models/Matter.php:
```diff
@@ -164,6 +190,15 @@ public function annuityAgents()
             ->wherePivot('role', 'ANN');
     }
 
+    public function sharedAnnuityAgents()
```
There must not be more than one, yet I don't recall this is checked. Adding more than one annuity agent, like more than one client, could have wrong side effects, so we'd need to design some safeguards (another project).

jjdejong commented on app/Models/Matter.php:
```diff
@@ -164,6 +190,15 @@ public function annuityAgents()
+            ->wherePivot('shared', 1);
+    }
+
     public function responsibles(): \Illuminate\Database\Eloquent\Relations\HasMany
```
There is only one responsible in a matter, specifically identified in the matter's `responsible` field.

jjdejong commented on app/Services/DocumentMergeService.php:
```diff
@@ -10,9 +10,18 @@
-    public function merge(Matter $matter, $filePath)
+    private Matter $matter;
+
+    public function setMatter(Matter $matter)
```
What does this bring? It appears to complicate readability by requiring lots of $this-> additions throughout the code.

AxelDeneu replied:
This approach eliminates the need to pass the Matter parameter repeatedly, which doesn’t make sense since these methods are private. They can’t be called externally from the service (or class), so using a property is more convenient. In my opinion, this also enhances readability, as it immediately conveys that the variable belongs to the object.

Additionally, using `setMatter()` avoids the need to initialize the Matter in the service constructor, as the service is already instantiated in the controller itself.

jjdejong commented on app/Services/DocumentMergeService.php:
So this, with the getXxxFields() methods, adds quite a few extra fields that are not documented . Is there a generic naming convention for the fields?

AxelDeneu replied:
I think the best naming convention would be to have fields like this :
- {Actore_Role_Name}
- {Actore_Role_Name}_Address
- {Actore_Role_Name}_Country
- etc.

Is it what you where asking for?

jjdejong replied:
OK for the new ones, but the old ones shouldn't break what we already have in all our letter templates! Is this the case?

AxelDeneu replied:
Nope that's not the case ;)

I’ve kept the old fields, which is why `getAgentFields` differs from the other `get{ActorRoleName}Fields` methods. This is because two fields already existed:

- Agent
- Agent_Ref

As for the Client, the way I build these fields remains unchanged. I’ve simply migrated the Client’s fields to a different location. The same applies to the Annuity Agent, with the addition of address fields for the Annuity Agent.

The fields for the other actors are entirely new.

jjdejong reviewed (commented):
I didn't understand the purpose of the setMatter() function, implying an extensive use of $this->matter instead of $matter. Some actors should be unique, even if the user can insert multiple ones. New merge fields require some documentation to be usable.

AxelDeneu committed:
fix(DocumentMergeService): Fixed relation being badly written, standardized hasOneThrough relation for actors depending on the role code

AxelDeneu committed:
fix(DocumentMergeService): Remove forgotten dd()

AxelDeneu commented:
First of all, I'm sorry you have to deal with this on a Sunday! ^^
That's a massive commit...

As you requested, I’ve fixed the issue with the relations returning multiple results when they should only return one. While I could have handled this by using the relation and simply returning the first result, I wasn’t comfortable with that approach since it bypasses eager loading. Instead, I created a method that allows us to use the pivot table as a `hasOneThrough` relation.

Additionally, I took the time to clean up the code and added some documentation.

Action Items :
- Could you please provide a list of other actor roles that should not have multiple results, so I can fix the relations accordingly?
- It might also be necessary to restrict the user to only adding one occurrence of these actors on the front end. Let me know if you’d like me to handle that as well.

Thanks again for your patience!

AxelDeneu committed:
fix(DocumentMergeService): Fix a bug not getting pivot table data from HasOneActorThroughActorPivot

jjdejong commented on app/Models/ActorPivot.php:
```diff
@@ -4,36 +4,70 @@
+    public function matter(): \Illuminate\Database\Eloquent\Relations\BelongsTo
     {
-        return $this->belongsTo(Matter::class);
+        return $this->belongsTo(Matter::class, 'matter_id');
```
Isn't `matter_id` Laravel's implicit key for that relationship? Why make it explicit?

jjdejong commented on app/Models/ActorPivot.php:
```diff
@@ -4,36 +4,70 @@
+    public function actor(): \Illuminate\Database\Eloquent\Relations\BelongsTo
     {
-        return $this->belongsTo(Actor::class);
+        return $this->belongsTo(Actor::class, 'actor_id');
```
Why make that key explicit?

jjdejong commented on app/Traits/BelongsToManyActors.php:
```diff
@@ -0,0 +1,40 @@
+            ->wherePivot('role', $role);
+
+        // If no relationship exists for the given role, return the container actors that are shared
+        if(!$request->exists()) {
```
How does this conditional $request definition play in terms of query building? This doesn't look SQL optimization friendly - it would generate two db queries. However, optimization is probably not crucial here because we are doing it for a single matter.

Apart from that, this should return all actors, those from the current matter plus those from the container matter, if shared. In SQL, this would need to be translated as a UNION ALL, so a ->merge() here.

jjdejong commented on app/Traits/HasOneActorThroughActorPivot.php:
```diff
@@ -0,0 +1,44 @@
+            ->where('role', $role);
+
+        // If no relationship exists for the given role, return the container actors that are shared
+        if(!$request->exists()) {
```
Same comment as above.

AxelDeneu replied:
You’re absolutely right—it’s not an ideal solution. I fully understand the concerns about using two queries. However, given the constraints of Laravel’s relation system, there isn’t a built-in way to perform a `UNION ALL` when defining relationships.

Using `->merge` would indeed disable eager loading, which means we’d have to call `->first()` on the final collection. That’s far from ideal when performance is a priority.

If this approach doesn’t work for you, I can refactor the code to use `->first()` for each actor’s role in the `actorsFromLnk` relation. However, this would result in the loss of eager loading, which is a significant downside in terms of performance.

Let me know how you’d like to proceed!

jjdejong replied:
Let's leave it like that for the document merging. I was thinking of using this to convert some raw SQL queries.

jjdejong replied:
This issue remains, though, which appears to need the use of a `->merge()` or a `->orWhere`:
> Apart from that, this should return all actors, those from the current matter plus those from the container matter, if shared. In SQL, this would need to be translated as a UNION ALL, so a ->merge() here.

AxelDeneu replied:
Yep, I will create an issue so I can provide you a new pull request

jjdejong reviewed (commented):
You're getting me lost with all this splitting up into multiple functions! But OK, if this is supposed to be "better". This is all for the document merging function, right? Or does it affect other functionality in the application?

jjdejong commented:
> * Could you please provide a list of other actor roles that should not have multiple results, so I can fix the relations accordingly?

I believe we have addressed them all.

> * It might also be necessary to restrict the user to only adding one occurrence of these actors on the front end. Let me know if you’d like me to handle that as well.

Yeah, I thought about this, but it's complicated with the sharing functionality. It hasn't really been a problem, by making sure only one actor is returned (using things like ->first()) where only one is expected.

AxelDeneu commented:
> You're getting me lost with all this splitting up into multiple functions! But OK, if this is supposed to be "better".

I know it can be difficult to follow when you're not used to splitting code into smaller pieces. But trust me, this is essentially the same issue, just in reverse. Reading through a file with around 800 lines is really challenging, and that’s why I’m taking extra time to contribute.

I understand that this might not be a problem for you, especially since you built the tool yourself and know it inside and out. But for me, and likely for other future contributors, it’s a different experience.

This is actually one of the core issues with open-source code: it should be maintainable and easy to navigate for anyone who comes across it.

---

> This is all for the document merging function, right? Or does it affect other functionality in the application?

These changes are specific to document merging. Before making any modifications to the files, I made sure to review other parts of the code that could potentially be affected. Additionally, a significant portion of the changes relates to methods I wrote myself for the document merging functionality.

---

> I believe we have addressed them all.

Perfect!

jjdejong commented:
Although I merged the PR, there are some unresolved conversations.

AxelDeneu commented:
Yep! Will create an issue for it
